kotest-framework/kotest-framework-engine/src/commonMain/kotlin/io/kotest/engine/test/interceptors/TimeoutInterceptor.kt

```diff
@@ -23,28 +38,79 @@ internal class TimeoutInterceptor(
       scope: TestScope,
       test: suspend (TestCase, TestScope) -> TestResult
    ): TestResult {
+      val timeout = testCase.config.timeout
 
-      // this timeout applies to the test itself. If the test has multiple invocations then
+      // This timeout applies to the test itself. If the test has multiple invocations, then
       // this timeout applies across all invocations. In other words, if a test has invocations = 3,
-      // each test takes 300ms, and a timeout of 800ms, this would fail, becauase 3 x 300 > 800.
-      logger.log { Pair(testCase.name.testName, "Switching context to add timeout ${testCase.config.timeout}") }
-
-      return when (val timeout = testCase.config.timeout) {
-         null -> test(testCase, scope)
-         else -> try {
-            withTimeout(timeout) {
-               test(testCase, scope.withCoroutineContext(coroutineContext))
-            }
-         } catch (t: Throwable) {
+      // each test takes 300ms, and a timeout of 800ms, this would fail, because 3 x 300 > 800.
+      logger.log { Pair(testCase.name.testName, "Switching context to add timeout $timeout") }
+
+      return try {
+         withAppropriateTimeout(timeout) {
+            test(testCase, scope.withCoroutineContext(coroutineContext))
+         }
+      } catch (t: CancellationException) {
+         if (t is WallclockTimeoutCancellationException || t is TimeoutCancellationException) {
             logger.log { Pair(testCase.name.testName, "Caught timeout $t") }
-            TestResult.Error(mark.elapsedNow(), TestTimeoutException(timeout, testCase.name.testName))
+            TestResult.Error(mark.elapsedNow(), TestTimeoutException(timeout, testCase.name.testName, t))
+         } else {
+            throw t
          }
       }
+
    }
 }
 
+// The implementation copied from Turbine:
+// https://github.com/cashapp/turbine/blob/1.1.0/src/commonMain/kotlin/app/cash/turbine/channel.kt#L93
+private suspend fun <T> withAppropriateTimeout(
+   timeout: Duration,
+   block: suspend CoroutineScope.() -> T,
+): T {
+   return if (coroutineContext[TestCoroutineScheduler] != null) {
+      // withTimeout uses virtual time, which will hang.
+      withWallclockTimeout(timeout, block)
+   } else {
+      withTimeout(timeout, block)
+   }
+}
+
+private suspend fun <T> withWallclockTimeout(
```

Comment:
this function is quite large. does it have its own unit tests?

Reply:
It is tricky to test it. I have only a test `ContainerTimeoutOverriddenMainDispatcherTest` failing without this function, but I'll try to write a separate test.